pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,565 @@
+		pub fn increment_or_create_allowance_count(
+			account_id: &T::AccountId,
+			currency_id: &T::CurrencyId,
+		) -> DispatchResult {
+			// not using try_mutate here as we're not sure if key exits, and we're already doing a some value check on result of exists query check
+			match <AccountCurrencyTransferRestriction<T>>::get(account_id, currency_id) {
+				Some(allowance_count) if allowance_count > 0 => {
+					let new_allowance_count = allowance_count
+						.checked_add(1)
+						.ok_or(Error::<T>::AllowanceCountOverflow)?;
+					<AccountCurrencyTransferRestriction<T>>::insert(
+						account_id,
+						currency_id,
+						new_allowance_count,
+					);
+					Ok(())
+				}
+				Some(_) => Err(DispatchError::from(
+					Error::<T>::AllowanceCountArithmeticError,
+				)),
+				_ => {
+					<AccountCurrencyTransferRestriction<T>>::insert(account_id, currency_id, 1);
+					Ok(())
+				}
+			}
+		}
+
+		/// Decrements the number of allowances tracked for a sending account/currency set.
+		/// If the allowance count is currently 1, then it removes the entry
+		/// If greater than 1, then decremented.
+		/// If no entry present, NoAllowancesSet error returned.
+		pub fn decrement_or_remove_allowance_count(
+			account_id: &T::AccountId,
+			currency_id: &T::CurrencyId,
+		) -> DispatchResult {
+			// not using try_mutate here as we're not sure if key exits, and we're already doing a some value check on result of exists query check
+			match <AccountCurrencyTransferRestriction<T>>::get(account_id, currency_id) {
+				Some(allowance_count) if allowance_count <= 1 => {
+					<AccountCurrencyTransferRestriction<T>>::remove(account_id, currency_id);
+					Ok(())
+				}
+				Some(allowance_count) => {
+					// check in this case should not ever be needed
+					let new_allowance_count = allowance_count
+						.checked_sub(1)
+						.ok_or(Error::<T>::AllowanceCountArithmeticError)?;
+					<AccountCurrencyTransferRestriction<T>>::insert(
+						account_id,
+						currency_id,
+						new_allowance_count,
+					);
+					Ok(())
+				}
+				_ => Err(DispatchError::from(Error::<T>::NoAllowancesSet)),
+			}
+		}
+	}
+
+	impl<T: Config> TransferAllowance<T::AccountId> for Pallet<T>
+	where
+		<T as frame_system::Config>::AccountId: Into<Location>,
+	{
+		type CurrencyId = T::CurrencyId;
+
+		fn allowance(
+			send: T::AccountId,
+			receive: Location,
+			currency: T::CurrencyId,
+		) -> Result<bool, DispatchError> {
+			match <AccountCurrencyTransferRestriction<T>>::get(&send, &currency) {
+				Some(count) if count > 0 => {
+					let current_block = <frame_system::Pallet<T>>::block_number();
+					match <AccountCurrencyTransferAllowance<T>>::get((&send, &currency, receive)) {
+						Some(AllowanceDetails {
+							allowed_at,
+							blocked_at,
+						}) if current_block >= allowed_at && current_block < blocked_at => Ok(true),
+						_ => Ok(false),
+					}
+				}
+				_ => Ok(true),
```

Comment:
is `count == 0` an error in the model, something that can not happen? If so, we are returning `Ok(true)` in that case. Should be an error instead?

Reply:
Ah, yeah I'll add a docstring & comments with the expected behaviour for the implementation to provide context.  It's currently working as intended--if there are 0 allowances added for the sending account and the currency then there shouldn't be any restrictions.

Reply:
👍 updated docs for impl